### PR TITLE
Scripts/Undercity: Journey to Undercity Scripted

### PR DESCRIPTION
**Pull request by jaredjones**
*Visual effects will now go off.
*Sylvannas will properly emote.
*Quest giver Ambassador Sunsorrow will now give you the lyrics of the song upon quest completion.
*SUnsorrow will also whisper you letting you know that he has something for you.

Closes: #12199
NOTE: This should also be cherry picked for 3.3.5a

### CODE REVIEW

File: sql/updates/world/2015_02_10_11_world.sql [hunk 1]
```diff
@@ -0,0 +1 @@
+UPDATE `world`.`creature_template` SET `ScriptName`='npc_ambassador_sunsorrow' WHERE `entry`='16287';
```
Comment (Kittnz) [inline sql/updates/world/2015_02_10_11_world.sql]: "`world`." should be removed

File: src/server/scripts/EasternKingdoms/zone_undercity.cpp [hunk 2]
```diff
@@ -69,27 +71,31 @@ float HighborneLoc[4][3]=
     {1292.51f, 310.50f, 1.99f},
 };
 
-#define HIGHBORNE_LOC_Y             -61.00f
-#define HIGHBORNE_LOC_Y_NEW         -55.50f
+#define HIGHBORNE_LOC_Y	                -61.00f
```
Comment (Kittnz) [inline src/server/scripts/EasternKingdoms/zone_undercity.cpp]: tab

**Comment by Kittnz** [comment]
Nice work :+1:

File: src/server/scripts/EasternKingdoms/zone_undercity.cpp [hunk 3]
```diff
@@ -239,18 +307,7 @@ class npc_highborne_lamenter : public CreatureScript
 
     struct npc_highborne_lamenterAI : public ScriptedAI
     {
-        npc_highborne_lamenterAI(Creature* creature) : ScriptedAI(creature)
-        {
-            Initialize();
```
Comment (DDuarte) [inline src/server/scripts/EasternKingdoms/zone_undercity.cpp]: This should not have been removed... variables will now be unitialized.

File: src/server/scripts/EasternKingdoms/zone_undercity.cpp [hunk 4]
```diff
@@ -157,15 +159,28 @@ class npc_lady_sylvanas_windrunner : public CreatureScript
...
                     DoSummon(ENTRY_HIGHBORNE_BUNNY, me, 10.0f, 3000, TEMPSUMMON_TIMED_DESPAWN);
-
+                    if (!flag) {
+                        Creature * sunsorrow = me->FindNearestCreature(AMBASSADOR_SUNSORROW, 40);
+                        if (sunsorrow && _player) {
+			    std::stringstream ss;
+			    ss << "That could have gone better ";
+			    ss << _player->GetName();
+			    ss << ", come here, please. I have something I need you to take care of for me.";
+			    sunsorrow->Whisper(ss.str(), LANG_UNIVERSAL,  _player);
```
Comment (DDuarte) [inline src/server/scripts/EasternKingdoms/zone_undercity.cpp]: This should use `broadcast_text` id 12157 and not a blatantly hack constructing the string by hand.

File: src/server/scripts/EasternKingdoms/zone_undercity.cpp [hunk 5]
```diff
@@ -99,23 +105,7 @@ class npc_lady_sylvanas_windrunner : public CreatureScript
 
     struct npc_lady_sylvanas_windrunnerAI : public ScriptedAI
     {
-        npc_lady_sylvanas_windrunnerAI(Creature* creature) : ScriptedAI(creature)
-        {
-            Initialize();
```
Comment (DDuarte) [inline src/server/scripts/EasternKingdoms/zone_undercity.cpp]: Again, do not remove the Initialize methods.

File: src/server/scripts/EasternKingdoms/zone_undercity.cpp [hunk 6]
```diff
@@ -69,27 +71,31 @@ float HighborneLoc[4][3]=
...
     npc_lady_sylvanas_windrunner() : CreatureScript("npc_lady_sylvanas_windrunner") { }
 
-    bool OnQuestReward(Player* /*player*/, Creature* creature, const Quest *_Quest, uint32 /*slot*/) override
+    bool OnQuestReward(Player* player, Creature* creature, const Quest *_Quest, uint32 /*slot*/) override
     {
         if (_Quest->GetQuestId() == QUEST_JOURNEY_TO_UNDERCITY)
         {
-            ENSURE_AI(npc_lady_sylvanas_windrunner::npc_lady_sylvanas_windrunnerAI, creature->AI())->LamentEvent = true;
-            ENSURE_AI(npc_lady_sylvanas_windrunner::npc_lady_sylvanas_windrunnerAI, creature->AI())->DoPlaySoundToSet(creature, SOUND_CREDIT);
+            CAST_AI(npc_lady_sylvanas_windrunner::npc_lady_sylvanas_windrunnerAI, creature->AI())->LamentEvent = true;
```
Comment (DDuarte) [inline src/server/scripts/EasternKingdoms/zone_undercity.cpp]: There's a reason these were changed to ENSURE_AI, please keep using ENSURE _AI.

File: src/server/scripts/EasternKingdoms/zone_undercity.cpp [hunk 7]
```diff
@@ -69,27 +71,31 @@ float HighborneLoc[4][3]=
     {1292.51f, 310.50f, 1.99f},
 };
 
-#define HIGHBORNE_LOC_Y             -61.00f
-#define HIGHBORNE_LOC_Y_NEW         -55.50f
+#define HIGHBORNE_LOC_Y	                -61.00f
+#define HIGHBORNE_LOC_Y_NEW             -55.50f
+#define GOSSIP_TEXT_GIVE_BOOK           "What will you give me?"
+#define GOSSIP_TEXT_EMOTE_INTRO         "Lady Sylvanas Windrunner looks down at the discarded necklace. In her sadness, the lady incants a glamour, which beckons forth Highborne spirits. The chamber resonates with their ancient song about the Sin'dorei...."
```
Comment (DDuarte) [inline src/server/scripts/EasternKingdoms/zone_undercity.cpp]: Place of gossip texts is the database.

File: src/server/scripts/EasternKingdoms/zone_undercity.cpp (hunk 6 above)
Comment (joschiwald) [inline src/server/scripts/EasternKingdoms/zone_undercity.cpp]: you could use UnitAI::sQuestReward and get rid of all these ai casts

**Comment by DDuarte** [comment]
> 16:42:54 \<Jared\> I knew they were gonna tear this PR apart. I rarely ever commit on TrinityCore.

You didn't return to IRC so I did not have the chance to talk about this.

By no means the comments we do about the code you submitted are meant to demoralize you from contributing and I'm sorry if it looks like that way. We do however have to do some filtering on code that gets added to Trinity for obvious reasons (keeping hacks out, maintainable/performant code, ...).

**Comment by jaredjones** [comment]
Oh I wasn't offended one bit. Every time I submit a PR Shauren usually rips me a new asshole. So all I was trying to say is that I was expecting this to happen. xD

File: src/server/scripts/EasternKingdoms/zone_undercity.cpp [hunk 8]
```diff
@@ -126,10 +116,23 @@ class npc_lady_sylvanas_windrunner : public CreatureScript
         uint32 BlackArrowTimer;
         uint32 ShotTimer;
         uint32 MultiShotTimer;
+        bool flag;
+        Player * _player;
```
Comment (jackpoz) [inline src/server/scripts/EasternKingdoms/zone_undercity.cpp]: Never store a Player pointer as class field in a script

**Comment by M0ra** [comment]
lol))

File: src/server/scripts/EasternKingdoms/zone_undercity.cpp (hunk 8 above)
Comment (jaredjones) [inline src/server/scripts/EasternKingdoms/zone_undercity.cpp]: I understand why this would typically be a bad thing to do; however only one person can trigger this script at a time so what's the problem?

**Comment by Unholychick** [comment]
please update to eventmap instead of maintaining outdated code.

**Comment by Nyeriah** [comment]
Sunsorrow's script could be done with SAI, just a suggestion :P it seems to be mostly just gossip scripting

File: src/server/scripts/EasternKingdoms/zone_undercity.cpp (hunk 8 above)
Comment (Nyeriah) [inline src/server/scripts/EasternKingdoms/zone_undercity.cpp]: We never really store pointers to items/players/creatures as they can be invalid at some point, and that may lead to unexpected outcomes. You can just store it's guid instead of you need to access it out of the local method scope. I see you're checking it against null but it's still better to keep code consistent

Comment (jackpoz) [inline src/server/scripts/EasternKingdoms/zone_undercity.cpp]: what happens if the player logs out after _player is set and before _player is checked ?

**Comment by jaredjones** [comment]
SAI is gross because it's hard to read IMO. Plus my knowledge of the DB is nil. If someone could write the broadcast_text SQL as well as the the text for sylvannas' gossip I'd be glad to add it to the PR and I'll remove the hard coded "C-Strings".

**Comment by jaredjones** [comment]
My personal opinion (not that it counts much) is that all SAI scripting should be changed to LUA in the future. This way it's both easier to read and yet doesn't require a recompile to the core whenever you change something.

**Comment by Nyeriah** [comment]
> SAI is gross because it's hard to read IMO.

I disagree, that's what comments are there for (and this script in special doesn't particularly seem to be over 4-5 lines of SAI). I also am not aware that it is in the plans to convert the current SAI scripts to LUA and future is pretty relative term... but that's also just my opinion

**Comment by jackpoz** [comment]
http://www.lua.org/about.html
>Please do not write it as "LUA", which is both ugly and confusing, because then it becomes an acronym with different meanings for different people. So, please, write "Lua" right!

**Comment by Rushor** [comment]
sniffing this soon for propper datas :)

**Comment by Rushor** [comment]
okay proff that thsi stuff is from retail:
http://pokit.org/get/?b9ab711b80c36161f2dee875a1cb2b95.jpg

the texts:
```sql
ServerToClient: SMSG_MESSAGECHAT (0x0603) Length: 176 ConnIdx: 0 EP:  Time: 02/12/2015 13:35:04.969 Number: 1157
Chat type: 15
Language: Universal (0)
SenderGUID: Full: 0x1C091000000FE7C000003200005AEBD5 Creature/0 R580/S50 Map: 0 Entry: 16287 Low: 5958613
SenderGuildGUID: Full: 0x0
WowAccountGUID: Full: 0x0
TargetGUID: Full: 0x081144000000000000000000061E6477 Player/0 R1105/S0 Map: 0 Low: 102655095
TargetVirtualAddress: 50528265
SenderVirtualAddress: 50593826
PartyGUID: Full: 0x0
AchievementID: 0
DisplayTime: 0
ChatFlags: 32
HideChatLog: False
FakeSenderName: False
Sender Name: Ambassador Sunsorrow
Receiver Name: 
Addon Message Prefix: 
Channel Name: 
Text: That could have gone better. $n, come here, please. I have something I need you to take care of for me.

ServerToClient: SMSG_GOSSIP_MESSAGE (0x1746) Length: 89 ConnIdx: 0 EP:  Time: 02/12/2015 13:35:59.688 Number: 3429
GossipGUID: Full: 0x1C091000000FE7C000003200005AEBD5 Creature/0 R580/S50 Map: 0 Entry: 16287 Low: 5958613
GossipID: 7178
FriendshipFactionID: 0
TextID: 8740
GossipOptions: 1
GossipText: 0
[0] ClientOption: 0
[0] OptionNPC: 0
[0] OptionFlags: 0
[0] OptionCost: 0
[0] Text: What is it that you have for me, ambassador?
[0] Confirm: 

```

if you need more pn me in irc :)

**Comment by jaredjones** [comment]
Thanks! I swear they've changed this text year over year. Can you tell me if you were playing on a Blood Elf or just a regular Horde character?

**Comment by Kittnz** [comment]
Blood elf as it shows in images.

**Comment by Aokromes** [comment]
/home/travis/build/TrinityCore/TrinityCore/src/server/scripts/EasternKingdoms/zone_undercity.cpp:268:59: fatal error: 
      use of undeclared identifier 'GOSSIP_TEXT_GIVE_BOOK'
                player->ADD_GOSSIP_ITEM(GOSSIP_ICON_TALK, GOSSIP_TEXT_GI...

**Comment by MitchesD** [comment]
whitespace everywhere

**Comment by jaredjones** [comment]
Whitespace removed.

File: src/server/scripts/EasternKingdoms/zone_undercity.cpp (hunk 3 above)
Comment (MitchesD) [inline src/server/scripts/EasternKingdoms/zone_undercity.cpp]: revert this change

**Comment by jaredjones** [comment]
Whitespace Removed

File: src/server/scripts/EasternKingdoms/zone_undercity.cpp [hunk 9]
```diff
@@ -69,26 +76,29 @@ float HighborneLoc[4][3]=
     {1292.51f, 310.50f, 1.99f},
 };
 
-#define HIGHBORNE_LOC_Y             -61.00f
-#define HIGHBORNE_LOC_Y_NEW         -55.50f
+#define HIGHBORNE_LOC_Y         -61.00f
+#define HIGHBORNE_LOC_Y_NEW     -55.50f
+#define GOSSIP_TEXT_GIVE_BOOK   "What will you give me?"
```
Comment (MitchesD) [inline src/server/scripts/EasternKingdoms/zone_undercity.cpp]: use DB for gossips

File: src/server/scripts/EasternKingdoms/zone_undercity.cpp [hunk 10]
```diff
@@ -224,8 +251,56 @@ class npc_lady_sylvanas_windrunner : public CreatureScript
...
+    }
+
+    bool OnGossipSelect(Player* player, Creature* creature, uint32 sender, uint32 action) override
+    {
+        switch (sender)
+        {
+            case GOSSIP_SENDER_INFO:
+                if (!player->GetItemCount(30632, true))
+                {
+                    player->AddItem(30632, 1);
```
Comment (MitchesD) [inline src/server/scripts/EasternKingdoms/zone_undercity.cpp]: use enum instead of magic number

File: src/server/scripts/EasternKingdoms/zone_undercity.cpp [hunk 11]
```diff
@@ -282,24 +350,25 @@ class npc_highborne_lamenter : public CreatureScript
...
 enum ParqualFintallas
 {
-    SPELL_MARK_OF_SHAME         = 6767
+    SPELL_MARK_OF_SHAME = 6767
 };
 
-#define GOSSIP_HPF1             "Gul'dan"
-#define GOSSIP_HPF2             "Kel'Thuzad"
-#define GOSSIP_HPF3             "Ner'zhul"
+#define GOSSIP_HPF1 "Gul'dan"
```
Comment (MitchesD) [inline src/server/scripts/EasternKingdoms/zone_undercity.cpp]: again, move gossips into database

File: src/server/scripts/EasternKingdoms/zone_undercity.cpp [hunk 12]
```diff
@@ -224,8 +251,56 @@ class npc_lady_sylvanas_windrunner : public CreatureScript
...
+    bool OnGossipSelect(Player* player, Creature* creature, uint32 sender, uint32 action) override
+    {
+        switch (sender)
+        {
+            case GOSSIP_SENDER_INFO:
+                if (!player->GetItemCount(30632, true))
+                {
+                    player->AddItem(30632, 1);
+                    player->CLOSE_GOSSIP_MENU();  /* close the menu */
+                    creature->Whisper("Here you are.", LANG_UNIVERSAL, player);
```
Comment (MitchesD) [inline src/server/scripts/EasternKingdoms/zone_undercity.cpp]: no

File: sql/updates/world/2015_03_08_11_world.sql [hunk 13]
```diff
@@ -0,0 +1,6 @@
+DELETE FROM creature_text WHERE entry=16287 AND groupid=0;
+DELETE FROM creature_text WHERE entry=10181 AND groupid=2;
+INSERT INTO creature_text (entry, groupid, id, text, type, language, probability, emote, duration, sound, BroadcastTextId, TextRange, comment) VALUES
```
Comment (MitchesD) [inline sql/updates/world/2015_03_08_11_world.sql]: to follow our coding standards creature_text -> `creature_text` , `entry` etc.

File: src/server/scripts/EasternKingdoms/zone_undercity.cpp [hunk 14]
```diff
@@ -17,17 +17,17 @@
  */
 
 /* ScriptData
```
Comment (Aokromes) [inline src/server/scripts/EasternKingdoms/zone_undercity.cpp]: On the rest of scriptdata we don't have spaces before comments

File: src/server/scripts/EasternKingdoms/zone_undercity.cpp [hunk 15]
```diff
@@ -17,17 +17,17 @@
...
-SDCategory: Undercity
-EndScriptData */
+ SDName: Undercity
+ SD%Complete: 95
+ SDComment: Quest support: 6628, 9180(post-event).
+ SDCategory: Undercity
+ EndScriptData */
 
 /* ContentData
-npc_lady_sylvanas_windrunner
```
Comment (Aokromes) [inline src/server/scripts/EasternKingdoms/zone_undercity.cpp]: On the rest of scriptdata we don't have spaces before comments

File: src/server/scripts/EasternKingdoms/zone_undercity.cpp [hunk 16]
```diff
@@ -224,8 +251,56 @@ class npc_lady_sylvanas_windrunner : public CreatureScript
 };
 
 /*######
```
Comment (Aokromes) [inline src/server/scripts/EasternKingdoms/zone_undercity.cpp]: On the rest of scriptdata we don't have spaces before comments

File: src/server/scripts/EasternKingdoms/zone_undercity.cpp [hunk 17]
```diff
@@ -224,8 +251,56 @@ class npc_lady_sylvanas_windrunner : public CreatureScript
...
+        return new npc_ambassador_sunsorrowAI(creature);
+    }
+
+    struct npc_ambassador_sunsorrowAI : public ScriptedAI
+    {
+        npc_ambassador_sunsorrowAI(Creature* creature) : ScriptedAI(creature) { }
+    };
+};
+
+/*######
```
Comment (Aokromes) [inline src/server/scripts/EasternKingdoms/zone_undercity.cpp]: On the rest of scriptdata we don't have spaces before comments

File: src/server/scripts/EasternKingdoms/zone_undercity.cpp [hunk 18]
```diff
@@ -282,24 +350,25 @@ class npc_highborne_lamenter : public CreatureScript
...
                     EventCast = false;
-                } else EventCastTimer -= diff;
+                }
+                else EventCastTimer -= diff;
             }
         }
     };
 };
 
 /*######
```
Comment (Aokromes) [inline src/server/scripts/EasternKingdoms/zone_undercity.cpp]: On the rest of scriptdata we don't have spaces before comments

**Comment by jaredjones** [comment]
Alright well I'm done working on this PR, feel free to close it or you can submit the 12 changes listed above onto my branch and I'll merge and rebase. I'm personally tired of working on this commit.

File: src/server/scripts/EasternKingdoms/zone_undercity.cpp (hunk 11 above)
Comment (Nyeriah) [inline src/server/scripts/EasternKingdoms/zone_undercity.cpp]: It's just a style change, that gossip doesn't belong to this PR

**Comment by Nyeriah** [comment]
```MySQL
DELETE FROM `gossip_menu_option` WHERE `menu_id` = 7178 AND `id`= 0;
INSERT INTO `gossip_menu_option` (`menu_id`, `id`, `option_icon`, `option_text`, `OptionBroadcastTextID`, `option_id`, `npc_option_npcflag`, `action_menu_id`, `action_poi_id`, `box_coded`, `box_money`, `box_text`, `BoxBroadcastTextId`) VALUES
(7178, 0, 0, 'What is it that you have for me, ambassador?', 19205, 1, 1, 0, 0, 0, 0, '', 0);

SELECT * FROM smart_scripts WHERE `entryorguid` IN (16287, 39372, 39371);
UPDATE `creature_template` SET `AIName` = 'SmartAI' WHERE `entry` = 16287;
DELETE FROM `smart_scripts` WHERE `entryorguid` = 16287 AND `source_type` = 0 AND `id` IN (0, 1);
INSERT INTO `smart_scripts` (`entryorguid`, `source_type`, `id`, `link`, `event_type`, `event_phase_mask`, `event_chance`, `event_flags`, `event_param1`, `event_param2`, `event_param3`, `event_param4`, `action_type`, `action_param1`, `action_param2`, `action_param3`, `action_param4`, `action_param5`, `action_param6`, `target_type`, `target_param1`, `target_param2`, `target_param3`, `target_x`, `target_y`, `target_z`, `target_o`, `comment`) VALUES
(16287,0,0,1,62,0,100,0,7178,0, 0,0,56,30632,1,0,0,0,0,7,0,0,0,0,0,0,0,'Ambassador Sunsorrow - On Gossip Option 0 Selected - Create Item \'Lament of the Highborn\''),
(16287,0,1,0,61,0,100,0,0,0,0,0,72,0,0,0,0,0,0,7,0,0,0,0,0,0,0,'Ambassador Sunsorrow - On Gossip Option 0 Selected - Close Gossip');
```

**Comment by Nyeriah** [comment]
By the way on a general note unrelated style changes make PRs in general harder to review so in my opinion these should be kept to minimal. Apart from these, there's not much left to be done in this PR other than a couple of notes and few other things.

I also see this is more or less the request of this https://github.com/TrinityCore/TrinityCore/issues/12199#issuecomment-44976409 but back at that time that script had several issues as well

File: src/server/scripts/EasternKingdoms/zone_undercity.cpp (hunk 9 above)
Comment (Nyeriah) [inline src/server/scripts/EasternKingdoms/zone_undercity.cpp]: It is wrong either way.

```SQL
DELETE FROM `gossip_menu_option` WHERE `menu_id` = 7178 AND `id`= 0;
INSERT INTO `gossip_menu_option` (`menu_id`, `id`, `option_icon`, `option_text`, `OptionBroadcastTextID`, `option_id`, `npc_option_npcflag`, `action_menu_id`, `action_poi_id`, `box_coded`, `box_money`, `box_text`, `BoxBroadcastTextId`) VALUES
(7178, 0, 0, 'What is it that you have for me, ambassador?', 19205, 1, 1, 0, 0, 0, 0, '', 0);
```

**Comment by Rushor** [comment]
http://pokit.org/get/?c75bb97d39122faf59890153cca0638d.jpg yep the whispering appears exactly 13 seconds after i turned in the quest. my old note was wrong, because my old proof was a video which was speeded up and so i thought it was instant :D me sry

**Comment by jaredjones** [comment]
I just did the quest last night on retail. You receive the message from Ambassador Sunsorrow right when she starts singing. Which is what happens in my script as well.

**Comment by Nyeriah** [comment]
Yeah I know that one should be there but I was asking about this one https://github.com/TrinityCore/TrinityCore/pull/14094/files#diff-5337184b3c0a901215f4dfd62f5655aaR283

In your script he whispers players when they take the book, but I'm unable to find such text in the broadcast texts and I don't see it in @Rushor 's screenshots.

**Comment by Nyeriah** [comment]
wowscrnshot_031615_195135

This also missing

**Comment by Killyana** [comment]
When you complete the quest by returning Sylvanas's necklace to her, the Blood Elf nearby whispers something to you when the event completes. When you go to him and ask him what he has for you, he gives you this book with the lyrics.
This can be done by SAI.

**Comment by Nyeriah** [comment]
> When you complete the quest by returning Sylvanas's necklace to her, the Blood Elf nearby whispers something to you when the event completes. When you go to him and ask him what he has for you, he gives you this book with the lyrics.
> This can be done by SAI.

Already done @ https://github.com/TrinityCore/TrinityCore/pull/14094#issuecomment-77803133

I just gotta add the npc text that shows up after picking the book, everything else's working already. I'll push once done with the database minor things around this.

**Comment by Killyana** [comment]
Yes I deleted my comment for this reason :)

**Comment by Killyana** [comment]
You need just to add conditions for the gossip text.
``` sql
DELETE FROM `npc_text` WHERE  `id` IN (10378, 8458);
INSERT INTO `npc_text` (`id`, `text0_0`, `text0_1`, `prob0`, `em0_1`, `BroadcastTextID0`) VALUES
(8458, "Greetings, $r.  I am Ambassador Sunsorrow of the sin'dorei, or blood elves as we are commonly known.  I have high hopes that this new bond between us and the Horde will be mutually beneficial.", "Greetings, $r.  I am Ambassador Sunsorrow of the sin'dorei, or blood elves as we are commonly known.  I have high hopes that this new bond between us and the Horde will be mutually beneficial.", 1, 2, 12161),
(10378, "Just a small songbook that I thought you might like to have.  It contains the lyrics to a song known as the Lament of the Highborne; the one that Lady Sylvanas glamoured from the air.", "", 1, 1, 19206);

DELETE FROM `gossip_menu` WHERE `entry`=7178 AND `text_id` IN (10378, 8458);
INSERT INTO `gossip_menu` (`entry`, `text_id`) VALUES 
(7178,10378),
(7178,8458);

**Comment by Nyeriah** [comment]
Thanks a lot @Killyana 

Spell 37099 has a very weird behavior when the trigger bunnies spawn on some positions, so most of time, the missile travels in slow motion. So I guess I'll leave that change out of this.

How it should look like: https://www.youtube.com/watch?v=daBYmCsaS3Y&feature=youtu.be
How it looks like: http://youtu.be/4ZtGYPmdTd0

It might indeed be the right spell, but it has issues that I don't really know how to solve at this time.

**Comment by Killyana** [comment]
Here some informations about the spell, it seems like it need a cpp script to adjust the coords of the missile.
http://pastebin.com/azGpcmyF
-- This is actually more like a heat-seeking missile
-- That always misses. 
-- However, if you use it with a Nearby-Circumference (Dest)
-- It looks awesome!

**Comment by jaredjones** [comment]
As you guys fix things submit them as a Pull Request on my branch so I can merge/rebase them into this PR.

**Comment by Nyeriah** [comment]
```SQL
DELETE FROM `creature_text` WHERE `entry` = 16287 AND `groupid`= 0;
DELETE FROM `creature_text` WHERE `entry` = 10181 AND `groupid` = 2;
INSERT INTO `creature_text` (`entry`, `groupid`, `id`, `text`, `type`, `language`, `probability`, `emote`, `duration`, `sound`, `BroadcastTextId`, `TextRange`, `comment`) VALUES
(16287, 0, 0, 'That could have gone better. $n, come here, please. I have something I need you to take care of for me.', 15, 0, 100, 0, 0, 0, 12157, 0, 'Ambassador Sunsorrow SAY_SUNSORROW_WHISPER'),
(10181, 2, 0, '%s looks down at the discarded necklace. In her sadness, the lady incants a glamour, which beckons forth Highborne spirits. The chamber resonates with their ancient song about the Sin''dorei...', 16, 0, 100, 0, 0, 0, 19197, 0, 'Lady Sylvanas Windrunner EMOTE_LAMENT');

DELETE FROM `gossip_menu_option` WHERE `menu_id` = 7178 AND `id`= 0;
INSERT INTO `gossip_menu_option` (`menu_id`, `id`, `option_icon`, `option_text`, `OptionBroadcastTextID`, `option_id`, `npc_option_npcflag`, `action_menu_id`, `action_poi_id`, `box_coded`, `box_money`, `box_text`, `BoxBroadcastTextId`) VALUES
(7178, 0, 0, 'What is it that you have for me, ambassador?', 19205, 1, 1, 7178, 0, 0, 0, '', 0);

UPDATE `creature_template` SET `AIName` = 'SmartAI' WHERE `entry` = 16287;
DELETE FROM `smart_scripts` WHERE `entryorguid` = 16287 AND `source_type` = 0 AND `id` = 0;
INSERT INTO `smart_scripts` (`entryorguid`, `source_type`, `id`, `link`, `event_type`, `event_phase_mask`, `event_chance`, `event_flags`, `event_param1`, `event_param2`, `event_param3`, `event_param4`, `action_type`, `action_param1`, `action_param2`, `action_param3`, `action_param4`, `action_param5`, `action_param6`, `target_type`, `target_param1`, `target_param2`, `target_param3`, `target_x`, `target_y`, `target_z`, `target_o`, `comment`) VALUES
(16287,0,0,0,62,0,100,0,7178,0, 0,0,56,30632,1,0,0,0,0,7,0,0,0,0,0,0,0,'Ambassador Sunsorrow - On Gossip Option 0 Selected - Create Item \'Lament of the Highborne\'');

SELECT * FROM conditions WHERE sourcetypeorreferenceid = 15 AND sourcegroup = 7178;

DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId` = 15 AND `SourceGroup` = 7178 AND `SourceEntry` = 0 AND `SourceId` = 0;
DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId` = 14 AND `SourceGroup` = 7178 AND `SourceEntry` = 10378 AND `SourceId` = 0;
INSERT INTO `conditions` (`SourceTypeOrReferenceId`, `SourceGroup`, `SourceEntry`, `SourceId`, `ElseGroup`, `ConditionTypeOrReference`, `ConditionTarget`, `ConditionValue1`, `ConditionValue2`, `ConditionValue3`, `NegativeCondition`, `ErrorType`, `ErrorTextId`, `ScriptName`, `Comment`) VALUES
(15, 7178, 0, 0, 0, 8, 0, 9180, 0, 0, 0, 0, 0, '', 'Ambassador Sunsorrow show gossip only if quest Journey to Undercity is rewarded'),
(15, 7178, 0, 0, 0, 16, 0, 512, 0, 0, 0, 0, 0, '', 'Ambassador Sunsorrow show gossip only if player is Blood Elf'),
(15, 7178, 0, 0, 0, 2, 0, 30632, 1, 0, 1, 0, 0, '', 'Ambassador Sunsorrow show gossip only if player doesn\'t have item Lament of Highborn'),
(14, 7178, 10378, 0, 0, 8, 0, 9180, 0, 0, 0, 0, 0, '', 'Ambassador Sunsorrow show gossip menu only if quest Journey to Undercity is rewarded');

SELECT * FROM conditions WHERE sourcetypeorreferenceid = 14 AND sourcegroup = 7188;

DELETE FROM `npc_text` WHERE  `id` IN (10378, 8458);
INSERT INTO `npc_text` (`id`, `text0_0`, `text0_1`, `prob0`, `em0_1`, `BroadcastTextID0`) VALUES
(8458, "Greetings, $r.  I am Ambassador Sunsorrow of the sin'dorei, or blood elves as we are commonly known.  I have high hopes that this new bond between us and the Horde will be mutually beneficial.", "Greetings, $r.  I am Ambassador Sunsorrow of the sin'dorei, or blood elves as we are commonly known.  I have high hopes that this new bond between us and the Horde will be mutually beneficial.", 1, 2, 12161),
(10378, "Just a small songbook that I thought you might like to have.  It contains the lyrics to a song known as the Lament of the Highborne; the one that Lady Sylvanas glamoured from the air.", "", 1, 1, 19206);

DELETE FROM `gossip_menu` WHERE `entry`=7178 AND `text_id` IN (10378, 8458);
INSERT INTO `gossip_menu` (`entry`, `text_id`) VALUES 
(7178,10378),
(7178,8458);

UPDATE `creature_text` SET `type` = 12 WHERE `entry` = 10181 AND `groupid` = 1;
```

I'm having problems dealing with that gossip that should only show when the gossip_menu_option is selected (to get the book) it happens to be always displayed on the npc's flavor text, any ideas on how to condition it to only show once the gossip is taken? Never really dealt with it (I dislike npc_texts and that sort of thing in general so never really explored it that far :P)

**Comment by jaredjones** [comment]
Sadly I'm not familiar with the TC database, especially the SAI part (gross.) The moment people were telling me to convert my beautiful easy to understand script into SAI/SQL I stopped working on it. xD

**Comment by Killyana** [comment]
According to the video there's another text, I will write the conditions later:
``` sql
DELETE FROM `gossip_menu` WHERE `entry`=7178 AND `text_id` IN (10378, 8458, 8740);
INSERT INTO `gossip_menu` (`entry`, `text_id`) VALUES 
(7178,10378),
(7178,8740),
(7178,8458);

**Comment by Nyeriah** [comment]
Yeah it uses a new text, but I was thinking it could be a cata change or something

**Comment by Killyana** [comment]
So which texts you want to use 10378 and 8740 ?

**Comment by Killyana** [comment]
Any way you can use 8458 or 8740 and when the player use gossip option the 10378 will be displayed, in this case we don't need conditions. 
``` sql
UPDATE `creature_template` SET `AIName` = 'SmartAI' WHERE `entry` = 16287;
DELETE FROM `smart_scripts` WHERE `entryorguid` = 16287 AND `source_type` = 0 AND `id` = 0;
INSERT INTO `smart_scripts` (`entryorguid`, `source_type`, `id`, `link`, `event_type`, `event_phase_mask`, `event_chance`, `event_flags`, `event_param1`, `event_param2`, `event_param3`, `event_param4`, `action_type`, `action_param1`, `action_param2`, `action_param3`, `action_param4`, `action_param5`, `action_param6`, `target_type`, `target_param1`, `target_param2`, `target_param3`, `target_x`, `target_y`, `target_z`, `target_o`, `comment`) VALUES
(16287,0,0,1,62,0,100,0,7178,0, 0,0,56,30632,1,0,0,0,0,7,0,0,0,0,0,0,0,'Ambassador Sunsorrow - On Gossip Option 0 Selected - Create Item \'Lament of the Highborne\''),
(16287,0,1,0,61,0,100,0,0,0, 0,0,98,7178,10378,0,0,0,0,7,0,0,0,0,0,0,0,'Ambassador Sunsorrow - On Gossip Option 0 Selected - SEND_GOSSIP_MENU'');

**Comment by Nyeriah** [comment]
Well it still shows the book gossip when I talk to him by default, so I guess conditions are really needed, but then again we fall on the same issue we were before
